beacon-chain/db/kv/blocks.go

```diff
@@ -238,6 +238,52 @@ func (s *Store) DeleteBlock(ctx context.Context, root [32]byte) error {
 	})
 }
 
+// DeleteBlocksAndStatesBeforeSlot performs deletes all blocks and states before the given slot.
+func (s *Store) DeleteBlocksAndStatesBeforeSlot(ctx context.Context, cutoffSlot primitives.Slot) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.DeleteBlocksAndStatesBeforeSlot")
+	defer span.End()
+
+	// Perform all deletions in a single transaction for atomicity
+	return s.db.Update(func(tx *bolt.Tx) error {
+		filter := filters.NewFilter().SetStartSlot(0).SetEndSlot(cutoffSlot)
+		_, rootsBySlot, err := blockRootsByFilter(ctx, tx, filter)
+		if err != nil {
+			return errors.Wrap(err, "could not retrieve block roots by filter")
+		}
+
+		for slot, roots := range rootsBySlot {
+			for _, root := range roots {
+				// Delete Block
+				if err = tx.Bucket(blocksBucket).Delete(root); err != nil {
+					return errors.Wrap(err, "could not delete block")
+				}
+
+				// Delete State
+				if err = tx.Bucket(stateBucket).Delete(root); err != nil {
+					return errors.Wrap(err, "could not delete state")
+				}
+
+				// Delete state summary
+				if err = tx.Bucket(stateSummaryBucket).Delete(root); err != nil {
+					return errors.Wrap(err, "could not delete state summary")
+				}
+
+				// Delete block from cache
+				s.blockCache.Del(string(root))
+				// Delete state summary from cache
+				s.stateSummaryCache.delete([32]byte(root))
```

Comment:
This might not be very important to fix, but if the transaction fails these deletions will not roll back. Maybe it would be better to store a slice of roots and then clear caches after the transaction completes? The downside of this is allocating the slice.